pennylane/fourier/reconstruct.py

```diff
@@ -0,0 +1,295 @@
+# Copyright 2018-2021 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Contains a function that computes the fourier series equivalent to
+a quantum expectation value."""
+from functools import wraps
+from inspect import signature
+import warnings
+
+# import numpy as np
+import pennylane as qml
+from pennylane import numpy as np
+
+
+def _reconstruct_equ(fun, num_frequency, fun_at_zero=None):
+    r"""Reconstruct a univariate Fourier series with consecutive integer
+    frequencies, using trigonometric interpolation and equidistant shifts.
+
+    This technique is based on
+    `Dirichlet kernels <https://en.wikipedia.org/wiki/Dirichlet_kernel>`_, see
+    `Vidal and Theis (2018) <https://arxiv.org/abs/1812.06323>`_ or
+    `Wierichs et al. (2021) <https://arxiv.org/abs/2107.12390>`_.
+
+    Args:
+        fun (callable): Function to reconstruct
+        num_frequency (int): Number of integer frequencies in ``fun``
+        fun_at_zero (float): Value of ``fun`` at zero; Providing ``fun_at_zero`` saves one
+            evaluation of ``fun``
+
+    Returns:
+        callable: Reconstructed Fourier series with ``num_frequency`` frequencies,
+        as ``qml.numpy`` based function.
+    """
+    if not abs(int(num_frequency)) == num_frequency:
+        raise ValueError(f"num_frequency must be a non-negative integer, got {num_frequency}")
+
+    a, b = (num_frequency + 0.5) / np.pi, 0.5 / np.pi
+    shifts_pos = qml.math.arange(1, num_frequency + 1) / a
+    shifts_neg = -shifts_pos[::-1]
+    fun_at_zero = fun(qml.math.array(0.0)) if fun_at_zero is None else fun_at_zero
+    evals = list(map(fun, shifts_neg)) + [fun_at_zero] + list(map(fun, shifts_pos))
```

Comment:
Nice! Quick thought: is `fun` always a QNode? if so, we could use [`batch_params`](https://pennylane.readthedocs.io/en/latest/code/api/pennylane.batch_params.html). The same number of evaluations are required, the only advantage of this is that the device could choose to parallelize those executions.

Reply:
Ah, this is a good idea. `fun` is going to be a lambda expression wrapping a QNode, does that work? :thinking:

Reply:
No unfortunately :( `@batch_params` needs the QNode object directly to extract info/attributes. we can always leave it then, and create an issue as a future enhancement

Reply:
Yeah, that wouldn't be something that's obvious to me to do..

Reply:
Created #1881 with an implementation idea :) @josh146